Asterix/WSC_functions.py

```diff
@@ -143,67 +253,273 @@ def creatingInterractionmatrix(input_wavefront, testbed, dimEstim,
+                                              wavefrontupstream)
+                if isinstance(
+                        OpticSysbefore,
+                        OptSy.deformable_mirror) and OpticSysbefore.active:
+                    # this subsystem is an active DM but not the one we actuate now (located before the one we actuate)
+
+                    if OpticSysbefore.z_position == 0:
+                        wavefrontupstream = wavefrontupstream * OpticSysbefore.EF_from_phase_and_ampl(
+                            phase_abb=OpticSysbefore.phase_init,
+                            wavelengths=wavelength)
+                    else:
+                        wavefrontupstream = OpticSysbefore.prop_pup_to_DM_and_back(
+                            wavefrontupstream, OpticSysbefore.phase_init,
+                            wavelength)
+
+                    if save_all_planes_to_fits == True:
+                        # save phase on this DM
+                        name_plane = 'Phase_init_on_' + osname + '_wl{}'.format(
+                            int(wavelength * 1e9))
+                        useful.save_plane_in_fits(dir_save_all_planes,
+                                                  name_plane,
+                                                  OpticSysbefore.phase_init)
+
+                else:
+                    wavefrontupstream = OpticSysbefore.EF_through(
+                        entrance_EF=wavefrontupstream)
+
+                if save_all_planes_to_fits == True:
+                    # save PP plane after this subsystem
+                    name_plane = 'EF_PP_after_' + osname + '_wl{}'.format(
+                        int(wavelength * 1e9))
+                    useful.save_plane_in_fits(dir_save_all_planes, name_plane,
+                                              wavefrontupstream)
+
+            # then the DM we want to actuate !
+            #
+            # if the DM is not in pupil plane, we can measure the first Fresnel transf only once
             if DM.z_position != 0:
+                wavefrontupstreaminDM, _ = prop.prop_fresnel(
+                    wavefrontupstream, DM.wavelength_0, DM.z_position,
+                    DM.diam_pup_in_m / 2, DM.prad)
+            
+            if visu:
+                plt.ion()
+                plt.figure()
 
-                #TODO make something smarter to check automatically the name of the entrance pup
-                Pup_inDMplane, _ = prop.prop_fresnel(testbed.entrancepupil.pup,
-                                                     DM.wavelength_0,
-                                                     DM.z_position,
-                                                     DM.diam_pup_in_m / 2,
-                                                     DM.prad)
+            # now we go throught the DM basis
+            init_pos_in_matrix = pos_in_matrix # where we store the next vect in the matrix
 
             for i in range(DM.basis_size):
 
                 if i % 10:
```

Comment:
ok si actuateur mais a du sens pour Fourier ?

Reply:
oui, ca fait la taille de la base, même si ne sait pas la quel c'est.

Le passage à la base se fait dans le fichier Optical_System_functions.py, les fonctions: 
- deformable_mirror.create_DM_basis(basis_type=basis_type) qui crée une base sauvée dans deformable_mirror.basis (2d numpy array of size [Size basis, Number of active act in the DM])

- testbed.basis_vector_to_act_vector, qui repasse d'un vecteur exprimé sur la base à un vecteur exprimé sur la base des actionneurs. 

Il pourrait être plus opportun de repasser cette seconde fonction dans la classe DM